vpr/SRC/route/route_timing.c

```diff
@@ -514,27 +561,110 @@ bool timing_driven_route_net(int inet, int itry, float pres_fac, float max_criti
-	int highfanout_rlim;
-
 	/* Rip-up any old routing. */
+	// convert the previous iteration's traceback into the starting route tree for this iteration
 
-	pathfinder_update_one_cost(trace_head[inet], -1, pres_fac);
-	free_traceback(inet);
-	
-	for (ipin = 1; ipin < g_clbs_nlist.net[inet].pins.size(); ipin++) {
-		if (!slacks) {
-			/* Use criticality of 1. This makes all nets critical.  Note: There is a big difference between setting pin criticality to 0
-			compared to 1.  If pin criticality is set to 0, then the current path delay is completely ignored during routing.  By setting
-			pin criticality to 1, the current path delay to the pin will always be considered and optimized for */
-			pin_criticality[ipin] = 1.0;
-		} else { 
+	INCREMENTAL_REROUTING_ASSERT(sanity_check_lookup());
+	// removing previous costs becomes the job of pruning afterwards
+
+	t_rt_node* rt_root;
+	remaining_targets.clear();	// efficient clearing by just resetting size without allocation
+	unsigned int num_sinks = g_clbs_nlist.net[inet].num_sinks();
+
+	if ((int)num_sinks < min_incremental_reroute_fanout || itry == 1) {
+#ifdef PROFILE
+		++entire_net_rerouted;
+#endif
+				pathfinder_update_one_cost(trace_head[inet], -1, pres_fac);
+		free_traceback(inet);
+
+		rt_root = init_route_tree_to_source(inet);
+		for (unsigned int sink_pin = 1; sink_pin <= num_sinks; ++sink_pin)
+			remaining_targets.push_back(sink_pin);
+		// don't have node information in this case, call old functions
+		mark_ends(inet);
+	}
+	else {
+		reached_sinks.clear();
+#ifdef PROFILE
+		clock_t rebuild_start = clock();
+#endif
+		// convert the previous iteration's traceback into the partial route tree for this iteration
+		rt_root = traceback_to_route_tree(inet);
+
+		// check for edge correctness
+		INCREMENTAL_REROUTING_ASSERT(is_valid_skeleton_tree(rt_root));
+		INCREMENTAL_REROUTING_ASSERT(should_route_net(inet));
+		INCREMENTAL_REROUTING_ASSERT(!is_uncongested_route_tree(rt_root));
+
+
+		bool destroyed {prune_route_tree(rt_root, pres_fac, reached_sinks, remaining_targets)};
+
+		// entire traceback is still freed since the pruned tree will need to have its pres_cost updated
+		pathfinder_update_one_cost(trace_head[inet], -1, pres_fac);
+		free_traceback(inet);
+
+		// if entirely pruned, have just the source (not removed from pathfinder costs)
+		if (destroyed) {
+#ifdef PROFILE
+			++entire_tree_pruned;
+			// traceback remains empty for just the root and no need to update pathfinder costs
+#endif
+		}
+		else {
+#ifdef PROFILE
+			++part_tree_preserved;
+#endif
+			// sync traceback into a state that matches the route tree
+			traceback_from_route_tree(inet, rt_root, num_sinks - remaining_targets.size());
+			// put the updated costs of the route tree nodes back into pathfinder
+			pathfinder_update_one_cost(trace_head[inet], 1, pres_fac);
+		}
+
+		// give lookup on the reached sinks
+		put_sink_rt_nodes_to_pins_lookup(rr_node_to_pin[inet], reached_sinks, rt_node_of_sink);

```

Comment:
Does rr_node_to_pin need to be a per-net datastructure (i.e. do we only query a single net at a time)?